arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Effect.kt

```diff
@@ -0,0 +1,242 @@
+ *
+ * So [Effect] is defined by [fold], to map both values of [R] and [A] to a value of `B`.
+ */
+public interface Effect<R, A> {
+  /**
+   * Runs the suspending computation by creating a [Continuation], and running the `fold` function
+   * over the computation.
+   *
+   * When the [Effect] has shifted with [R] it will [recover] the shifted value to [B], and when it
+   * ran the computation to completion it will [transform] the value [A] to [B].
+   *
+   * ```kotlin
+   * import arrow.core.continuations.effect
+   * import io.kotest.matchers.shouldBe
+   *
+   * suspend fun main() {
+   *   val shift = effect<String, Int> {
+   *     shift("Hello, World!")
+   *   }.fold({ str: String -> str }, { int -> int.toString() })
+   *   shift shouldBe "Hello, World!"
+   *
+   *   val res = effect<String, Int> {
+   *     1000
+   *   }.fold({ str: String -> str.length }, { int -> int })
+   *   res shouldBe 1000
+   * }
+   * ```
+   * <!--- KNIT example-effect-01.kt -->
+   */
+  public suspend fun <B> fold(
+    recover: suspend (shifted: R) -> B,
+    transform: suspend (value: A) -> B
+  ): B
+
+  /**
+   * Like `fold` but also allows folding over any unexpected [Throwable] that might have occurred.
+   * @see fold
+   */
+  public suspend fun <B> fold(
+    error: suspend (error: Throwable) -> B,
+    recover: suspend (shifted: R) -> B,
+    transform: suspend (value: A) -> B
+  ): B =
+    try {
+      fold(recover, transform)
+    } catch (e: Throwable) {
+      error(e.nonFatalOrThrow())
+    }
+
+  /**
+   * [fold] the [Effect] into an [Either]. Where the shifted value [R] is mapped to [Either.Left], and
+   * result value [A] is mapped to [Either.Right].
+   */
+  public suspend fun toEither(): Either<R, A> = fold({ Either.Left(it) }) { Either.Right(it) }
+
+  /**
+   * [fold] the [Effect] into an [Ior]. Where the shifted value [R] is mapped to [Ior.Left], and
+   * result value [A] is mapped to [Ior.Right].
+   */
+  public suspend fun toIor(): Ior<R, A> = fold({ Ior.Left(it) }) { Ior.Right(it) }
+
+  /**
+   * [fold] the [Effect] into an [Validated]. Where the shifted value [R] is mapped to
+   * [Validated.Invalid], and result value [A] is mapped to [Validated.Valid].
+   */
+  public suspend fun toValidated(): Validated<R, A> =
+    fold({ Validated.Invalid(it) }) { Validated.Valid(it) }
+
+  /**
+   * [fold] the [Effect] into an [Option]. Where the shifted value [R] is mapped to [Option] by the
+   * provided function [orElse], and result value [A] is mapped to [Some].
+   */
+  public suspend fun toOption(orElse: suspend (R) -> Option<A>): Option<A> = fold(orElse, ::Some)
+
+  /**
+   * [fold] the [Effect] into an [A?]. Where the shifted value [R] is mapped to
+   * [null], and result value [A].
+   */
+  public suspend fun orNull(): A? = fold({ null }, ::identity)
+
+  /** Runs the [Effect] and captures any [NonFatal] exception into [Result]. */
+  public suspend fun attempt(): Effect<R, Result<A>> = effect {
```

Comment:
```suggestion
  public fun attempt(): Effect<R, Result<A>> = effect {
```

All functions that return `Effect` should be non-suspending.